docs/testing-with-django.md

```diff
@@ -1,8 +1,237 @@
+appproach of calling the function and inspecting the results.
+
+For these types of tests, you will often use methods like
+`self.assertEqual()`, `self.assertTrue()`, etc., which come with
+Django.
+
+### Fixture-driven tests
+
+Particularly for testing Zulip's integrations with third party systems,
+we strive to have a highly data-driven approach to testing.  To give a
+specific example, when we test our GitHub integration, the test code
+reads a bunch of sample inputs from a JSON fixture file, feeds them
+to our Github integration code, and then verifies the output against
+expected values from the same JSON fixture file.
+
+Our fixtures live in `zerver/fixtures`. 
+
+### Mocks and stubs
+
+We use mocks and stubs for all the typical reasons:
+
+- to more precisely test the target code
+- to stub out calls to third-party services
+- to make it so that you can run your tests on the airplane without wifi
+
+For mocking we generally use the "mock" library and use `mock.patch` as
+a context manager.  We also take advantage of some context managers
+from Django as well as our own custom helpers.  Here is an example:
+
+
+        with self.settings(RATE_LIMITING=True):
+            with mock.patch('zerver.decorator.rate_limit_user') as rate_limit_mock:
+                api_result = my_webhook(request)
+
+        self.assertTrue(rate_limit_mock.called)
+
+### Template tests
+
+In [zerver/tests/test_templates.py](https://github.com/zulip/zulip/blob/master/zerver/tests/test_templates.py)
+we have a test that renders all of our back end templates with
+a "dummy" context, to make sure the templates don't have obvious
+errors.  (These tests won't catch all types of errors; they are
+just a first line of defense.)
+
+### SQL performance tests
+
+A common class of bug with Django systems is to handle bulk data in
+an inefficient way, where the back end populates objects for join tables
+with a series of individual queries that give O(N) latency.  (The
+remedy is often just to call `select_related()`, but sometimes it
+requires a more subtle restructuring of the code.)
+
+We try to prevent these bugs in our tests by using a context manager
+called `queries_captured()` that captures the SQL queries used by
+the back end during a particular operation.  We make assertions about
+those queries, often simply asserting that the number of queries is
+below some threshold.
+
+### Event-based tests
+
+The Zulip back end has a mechanism where it will fetch initial data
+for a client from the database, and then it will subsequently apply
+some queued up events to that data to the data structure before notifying
+the client.  The `EventsRegisterTest.do_test()` helper helps tests
+verify that the application of those events via apply_events() produces
+the same data structure as performing an action that generates said event.
+(This is a bit esoteric, but if you read the tests, you will see some of
+the patterns.)
```

Comment:
I think we have some docs on this part of the testing system in the new feature tutorial that we should cross-link with.